File: PHP/CommissionEmployee/Transportation-Management-System.php
<?php
abstract class Vehicle {
    protected $vehicleId;
    protected $model;
    protected $fuelLevel;

    public function __construct($vehicleId, $model, $fuelLevel) {
        $this->vehicleId = $vehicleId;
        $this->model = $model;
        $this->fuelLevel = $fuelLevel;
    }

    public function refuel($liters) {
        $this->fuelLevel += $liters;
        echo "$this->model refueled. New fuel level: $this->fuelLevel liters\n";
    }

    abstract public function calculateRange();
}

class Car extends Vehicle {
    private $fuelEfficiency;

    public function __construct($vehicleId, $model, $fuelLevel, $fuelEfficiency) {
        parent::__construct($vehicleId, $model, $fuelLevel);
        // since effeciency is measured in percentage, lets make sure we dont pass in a negative value
        if( $fuelEfficiency < 0.0){
            throw new InvalidArgumentException("fuel efficiency cannot be negative.");
        }
        $this->fuelEfficiency = $fuelEfficiency;
    }

    public function calculateRange() {
        $range = $this->fuelLevel * $this->fuelEfficiency;
        echo "$this->model range: $range km\n";
        return $range;
    }
}

class TransportationManager {
    public function operateVehicle(Vehicle $vehicle) {
        $vehicle->calculateRange();
    }

    // Compile-time polymorphism (method overloading through default arguments)
    public function refuelVehicle(Vehicle $vehicle, $liters, $fuelType = null) {
        if ($fuelType) {
            echo "Refueling {$vehicle->model} with $fuelType\n";
        }
        $vehicle->refuel($liters);
    }
}

// Implementation Tasks
// A. Create Vehicles:
// Car("C001", "Sedan", 50, 15)
// B. Refuel Sedan 
// C. Use operateVehicle() to process all

$solomonCar= new Car("C001", "Sedan", 50, 15);
$solomonCar->refuel(20);
$manager = new TransportationManager();
$manager->operateVehicle($solomonCar);

?>

File: python_lab_tasks/TransportationManagementSystem.py
from abc import ABC, abstractmethod
class Vehicle(ABC):
    def __init__(self, vehicle_id, model, fuel_level):
        self.vehicle_id = vehicle_id
        self.model = model
        self.fuel_level = fuel_level

    def refuel(self, liters):
        self.fuel_level + liters
    
    @abstractmethod
    def calculate_range(self):
        pass

    def display_vehicle_details(self):
        return (f"Vehicle ID: {self.vehicle_id}\n"
                f"Model: {self.model}\n"
                f"Fuel Level: {self.fuel_level} liters")

class Car(Vehicle):
    def __init__(self, vehicle_id, model, fuel_level, fuel_efficiency):
        super().__init__(vehicle_id, model, fuel_level)
        self.fuel_efficiency = fuel_efficiency

    def calculate_range(self):
        range_km = self.fuel_level * self.fuel_efficiency
        print(f"Car {self.model} can travel {range_km} km with current fuel level.")

    def display_vehicle_details(self):
        return (super().display_vehicle_details() +
                f"\nFuel Efficiency: {self.fuel_efficiency} km/l")

class TransportationManager:
    @staticmethod
    def operate_vehicle(vehicle):
        vehicle.calculate_range()

# Implementation Tasks
if __name__ == "__main__":
    car = Car("C001", "Sedan", 50, 15)
    print(car.display_vehicle_details())
    car.refuel(10) 
    manager = TransportationManager()
    manager.operate_vehicle(car)